examples/Nat/NAT.cs

```diff
@@ -0,0 +1,148 @@
+                                 ---------
+
+It maps TCP connections arriving on non-zero ports, onto the NAT's zero port.
+It maintains a mapping of active TCP connections. It adds an entry to this mapping
+whenever it receives a TCP SYN packet coming from a non-zero port; it assigns a fresh
+ephemeral TCP port on its zero network port, and updates the source IP address to its
+own. It then forwards the packet onto its zero port. It removes an entry when it's no
+longer needed (because of a FIN-shutdown and timeout) or because an activity timer has
+expired. As long as it has an entry, it maps TCP segments from non-zero to zero ports,
+masquerading the client.
+Currently RST packets do not cause entries to be removed. Instead, the inactivity
+resulting from the connection being closed eventually triggers the removal.
+UDP connections are treated in a similar manner, but any UDP packet from Inside can
+cause an entry to be added where there wasn't one before.
+
+NOTE could improve the implementation by having configurable "forwarding ports"
+     that enable you to run servers on non-zero network ports.
+*/
+
+using System;
+using System.Net;
+using System.Net.NetworkInformation;
+using System.Timers;
+using PacketDotNet;
+
+namespace Pax.Examples.Nat
+{
+  /// <summary>
+  /// A packet processor that performs Network Address Translation (a NAT).
+  /// </summary>
+  public sealed class NAT : SimplePacketProcessor
+  {
+    /// A value indicating the packet should be dropped.
+    public const int Port_Drop = -1;
+
+    private readonly Timer gcTimer;
+
+    // Use a separate namespace for each transport protocol
+    private TcpNAT tcpNat;
+    private UdpNAT udpNat;
+
+    /// <summary>
+    /// Creates a new NAT packet processor that handles TCP and UDP packets.
+    /// </summary>
+    /// <param name="my_address">The public IP address of the NAT</param>
+    /// <param name="next_outside_hop_mac">The MAC address of the next hop on the outside-facing port.</param>
+    /// <param name="tcp_inactivity_timeout">The time that should elapse before a TCP connection with no activity is removed.</param>
+    /// <param name="tcp_time_wait_duration">The time that should elapse before a closed TCP connection is removed.</param>
+    /// <param name="tcp_start_port">The start of the range of TCP ports to use (inclusive).</param>
+    /// <param name="tcp_end_port">The end of the range of TCP ports to use (inclusive).</param>
+    /// <param name="udp_inactivity_timeout">The time that should elapse before a UDP connection with no activity is removed.</param>
+    /// <param name="udp_start_port">The start of the range of UDP ports to use (inclusive).</param>
+    /// <param name="udp_end_port">The end of the range of UDP ports to use (inclusive).</param>
+    /// <remarks>
+    /// The constructor parameters can be specified in the wiring configuration file, in the `args` dictionary of the handler entry.
```

Comment:
I very much like how you arranged for these parameters to be cleanly (and type-safely) extracted from the config (and allowed for optional parameters, and the support for multiple constructors in a clean way).